### PR TITLE
feat(nodes): save thumbnails

### PR DESCRIPTION
**Pull request by psychedelicious**
Save and serve thumbnail images

### CODE REVIEW

**Review by blessedcoolant — APPROVED** [review]

**Comment by psychedelicious** [comment]
This needs @Kyle0654 's approval, I'm not sure if I've done this correctly

File: invokeai/app/util/save_thumbnail.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    image: Image.Image,
+    filename: str,
+    path: str,
+    size: int = 256,
+) -> str:
+    """
+    Saves a thumbnail of an image, returning its path.
+    """
+    base_filename = os.path.splitext(filename)[0]
+    thumbnail_path = os.path.join(path, base_filename + ".webp")
```
Comment (Kyle0654) [inline invokeai/app/util/save_thumbnail.py:15]: Why are these webp?

File: invokeai/app/util/save_thumbnail.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    if os.path.exists(thumbnail_path):
+        return thumbnail_path
+
+    thumbnail_width = size
+    thumbnail_height = round(size * (image.height / image.width))
+
+    image_copy = image.copy()
+    image_copy.thumbnail(size=(thumbnail_width, thumbnail_height))
```
Comment (Kyle0654) [inline invokeai/app/util/save_thumbnail.py:24]: You don't need to pre-calculate width/height for the PIL thumbnail function. You just pass it the maximum size (e.g. 128x128) and it'll calculate the thumbnail size appropriately to maintain the aspect ratio.

**Review by Kyle0654 — APPROVED** [review]

File: invokeai/app/util/save_thumbnail.py (hunk 1 above)
Comment (blessedcoolant) [reply invokeai/app/util/save_thumbnail.py:15]: Small file size. Loading a ton of jpeg / png files in the gallery as thumbnails can get resource intensive. So we shifted to using webp's for the thumbnails in the gallery.

Comment (psychedelicious) [reply invokeai/app/util/save_thumbnail.py:15]: For reference, 512x512 png result vs 256x256 thumbnail:

```bash
520K    fea6cf48-ad19-4d31-a583-e36879645773_449210a3-5243-4b6c-abf2-c65458a56220_1680611408.png
20K     thumbnails/fea6cf48-ad19-4d31-a583-e36879645773_449210a3-5243-4b6c-abf2-c65458a56220_1680611408.webp
```

**Pull request merged**